### PR TITLE
Add/large bundle

### PR DESCRIPTION
**Pull request by iuravic**
This contains:
- a suggested step-by-step feature testing
- a description of the "conversion batch" concept which is being used,
- code layout description (to ease reviewing of this this unforgivably large PR),
- instructions on how to run the PHP test suite (optional).

#### Test data
Before installing and Activating the plugin, prepare a test DB, with some test data for conversion and patching.

This SQL file contains a sample of 1000 `wp_posts` from one of our publishers (file not publicly available):
https://drive.google.com/file/d/1xPNoNbOu-bwQl38_xll3mQd-ydCLb4lt/view?usp=sharing
- truncate your `wp_posts` table, then import the contents of this SQL file.

#### Build
- clone this repo, and switch to this branch
- run `composer install`
- `nvm use 10.10.0 && npm ci && npm run clean && npm run build:webpack -- --watch`

## 1. Plugin Installation
- go to WP-admin and click "Activate" the plugin,
- to confirm successful installation, you could check whether the DB was updated, or simply continue testing:
  - `SELECT * FROM wp_options WHERE option_name LIKE 'ncc%';` -- 6 initial options
  - `SELECT COUNT(*) AS total FROM ncc_wp_posts;` -- new table, should contain 1000 entries

## 2. The settings page.

In WP-admin's menu, click the the first link "Newspack Content Converter".

This opens the settings page. See that all the params are displaying in their fields. The Settings page is currently just an overview, and it doesn't enable changing any of the params.

At this point, please note that the whole interface (especially the Settings page you're looking at) is a dev tool, neither polished, nor made to be used by end users. For now this has an UI of a dev tool, made to enable on-site conversion and content re-patching.

## 3. The content conversion

To run content conversion, click the "Run conversion" menu link in the WP-admin. The page displays a conversion plan (currently 1000 posts, with 10 conversion batches, converting 100 posts each).

Click "Start Conversion Now" button to start converting. Conversion of the first batch of content should start. While the conversion is running in a tab, now proceed to manually open several other tabs (max 4 tabs recommended) and manually enter the Converter app's address: `/wp-admin/post-new.php?newspack-content-converter` . Just make sure not to disturb the active tab.

You can see a more detailed progress in the **console logs**.

Each new tab picks up the a batch and converts it. Here's how this might look like when using 4:

image

Now hold for a little while while the conversion runs. It took 12 mins to convert this content on my computer.

If the browser pops up an _"Are you sure you wish to Reload the page?"_ click yes. I attempted to remove this popup entirely programmatically ([here](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-84f99dbb6cee27334b28665fb6c489e8R46), and [here](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-e70b04386946903cb7c87b89a6de7cc0R40)), but sometimes it might still show up.

Once completed, close the tabs, and Posts sections, make sure that they've been converted to blocks.

At this point, all the content patchers have been applied, too. 👍 

## 4. Re-applying patchers
From here we can apply patchers on top of the already converted content (e.g. you pull the repo containing some new patchers, then apply those).

Patching runs much faster than the conversion, and it runs in a similar way -- you may use multiple tabs to execute it, but the execution will be so fast, you shouldn't attempt to do it -- simply test it running in one tab.

To start patching, go to the "Re-apply patchers *dev*" menu in WP-admin. Note the patching plan outline, similar to the conversion plan in the previous step. When ready, click the "Patch content now" button.

## 5. Uninstalling the plugin
Note that this *will delete all the plugin files*, so if you wish to keep the plugin locally, you may wish to quickly `cp` the contents to a safe place, then bring them back after uninstallation.

Head on to the plugin page. First click "Deactivate", and then click "Delete".

Check that all the options have been removed:
- `SELECT * FROM wp_options WHERE option_name LIKE 'ncc%';`
and that the table is no longer there:
- `SELECT COUNT(*) AS total FROM ncc_wp_posts;`

## Explanation of the "conversion batches"

Since the conversion takes place in the frontend (React app integrates on top of the Block Editor, manipulates it to sequentially convert entries) it depends on a browser tab being kept active. Two points were relevant to deciding to use conversion in batches:
- using multiple browser tabs to run the app at the same time leads to a faster conversion,
- one more behavior was noticed - if the the Block Editor was used for a longer while, the conversion becomes a bit slower; refreshing it would speed it up to max again.

For these reasons the plugin divides the content in batches, and an auto-reload of the tab upon completion picks up the next batch.

Just to note, the (re-)patching also uses the concept of batches, but it works in the backend -- the fronten sends an API request to run the next batch, then awaits for completion response.

### Next step with this plugin

... should probably be to replace the use of multiple tabs with multiple iframes inside a single page. I'd predict the iframes might work out-of-the-box, since the batch processing has been made to support such use.

--------------------------------

### Code layout

Frontend:
`assets/src/`:
- [`index.js`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-e70b04386946903cb7c87b89a6de7cc0) - all the `render()`s are called from here
- [`style.css`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-603658affcb1425f16b5dd8c26c47583) - a simple base .css for all the pages (since they're so basic and dev-minded right now)
- [`content-converter/`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-84f99dbb6cee27334b28665fb6c489e8) - the conversion app, sitting on top of Gutenberg
- [`content-repatcher/`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-ee30aee1659c52aaf9b719ffe06f9ea8) - the part that runs the content patchers
- [`conversion/`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-f72953981e26d1c79101da50fe6734f0) - the page with the "Convert now" button
- [`patchers/`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-baa031ccec4a1bc3ffc9b96827055b4d) - the page with the "Patch now" button
- [`settings/`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-a4273a1038442de773788f3372a46c15) - the settings page
- [`utilities/`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-242c0aaf2b3b6c54d382fc62a760dd39) - this file got new API calls

**Plugin files:**
[`newspack-content-converter.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-3e6d6def1317cc875be26a831ad90cb8) - the plugin file
[`dependency-includer-script.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-d720ff5758637b1dc8da4b6222ed6219) - script which includes all the dependencies
`lib/`:
- [`lib/class-config.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-4a5de49daebe51788ee7c46e6d1d647e) - a centralized place to keep track of all the options and params (not to have them mix around, ignored, repeated or confused)
- [`lib/class-installer.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-81e6a341b8a091cb2e2b44d41d4586bd) - plugin installer class
- [`uninstall.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-96bb271b1c1e165827bfcf09d4961e80) - this is a magic file in WP, better than using the uninstall hook because when it gets run, the `WP_UNINSTALL_PLUGIN` is defined, unlike the uninstall hook
- [`lib/class-converter.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-a5b02d72af040acf5047fedf5165521e) - main plugin class, now mostly plugin config and API calls
- [`lib/class-conversionprocessor.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-c34fbbef97f936a54b104895827ebfe9) - does all the conversion and patching

**The patchers:** actual content patchers
[`lib/content-patcher/class-patchhandler.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-2ddda7ec0fb751b650899daa9919f4c3)
- [`lib/content-patcher/patchers/class-imgpatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-a01876c82697a956d8f5baa383775cc3) -- the most complicated of all so far, it could use refactoring/simplifying
- [`lib/content-patcher/patchers/class-patcherabstract.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-e79fbb41b00766883f8691f3f2225d12)
- [`lib/content-patcher/patchers/class-audiopatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-ccc36cd165ef8941dcfd8ed8fba3da6e)
- [`lib/content-patcher/patchers/class-blockquotepatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-62bac135a2dba65b4a6909668033e7e6)
- [`lib/content-patcher/patchers/class-captionimgpatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-69df63fec59e7058faafaf53d85a713e)
- [`lib/content-patcher/patchers/class-paragraphpatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-fda51607be44f50b9fe205d9db60f112)
- [`lib/content-patcher/patchers/class-videopatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-18ac2fe6b0f08f3b0823162297ff79af)

**Element manipulators:**
Here is where all the element logic got put, all the "regex stuff", or rather all the logic that manipulates the HTML and block elements. From this place, the htmlelementmanipulator can easily be refactored to use the DOM library (and running the tests will make sure nothing gets broken):
`lib/content-patcher/elementManipulators/`
- [`class-htmlelementmanipulator.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-62cdc9e9632ba2b7d44e92f3216e84ec) -- this one's a candidate to be replaced by a DOM manipulating library
- [`class-squarebracketselementmanipulator.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-f4c9beab6b02508e7c3af83c4de3f28c)
- [`class-wpblockmanipulator.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-dd33af78f85ba501e3a852bb536c288a)

The test suite -- they contain > 70 real-life cases, and running them makes sure none of the base functionality gets broken. Quite valuable by now. Reading the method names shows what we're testing for.
[`phpunit.xml.dist`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-98c12a693941ae3c4f50872403416471)
- [`tests/unit/content-patcher/patchers/test-img-patcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-a1fb0849829c7d5798b1d222bb8e7162)
- [`tests/unit/content-patcher/patchers/test-audio-patcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-c9dbc36bb2f2972bf2dc2ce55eeca384)
- [`tests/unit/content-patcher/patchers/test-blockquote-patcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-26c4f71a7926ac59547ee7117d841340)
- [`tests/unit/content-patcher/patchers/test-caption-img-patcher.php`](tests/unit/content-patcher/patchers/test-caption-img-patcher.php)
- [`tests/unit/content-patcher/patchers/test-paragraph-patcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-8780d05b9856639a8a58230254ca6a2f)
- [`tests/unit/content-patcher/patchers/test-video-patcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-c68fcb26bbb7153dad06c595ab0554ee)
These contain test data, both real cases (sanitized), and model cases:
- [`tests/fixtures/unit/content-patcher/patchers/class-dataproviderimgpatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-3e29fac6a3e548a71648382879bb4600)
- [`tests/fixtures/unit/content-patcher/patchers/class-dataprovideraudiopatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-b7f590a4bba8a90ae2bc5082e20257dc)
- [`tests/fixtures/unit/content-patcher/patchers/class-dataproviderblockquotepatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-b3d0033a926c2186b53bd7303161ad79)
- [`tests/fixtures/unit/content-patcher/patchers/class-dataprovidercaptionimgpatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-533c63d801d6a1d6a775567d660ed3f1)
- [`tests/fixtures/unit/content-patcher/patchers/class-dataproviderparagraphpatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-b523eeffcb2311184155146f7af6349f)
- [`tests/fixtures/unit/content-patcher/patchers/class-dataprovidervideopatcher.php`](https://github.com/Automattic/newspack-content-converter/pull/12/files#diff-e23b2d996c5e7d7724b443b495eb3e3c)

### Running the tests

The `phpunit` is not currently included in `composer`, so if you have it installed globally, simply run:
```
$ phpunit
```
Otherwise this will do the trick:
```
$ wget -O phpunit https://phar.phpunit.de/phpunit-7.phar
$ chmod +x phpunit
$ ./phpunit
```

### CODE REVIEW

File: assets/src/settings/index.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
+/**
```
Comment (jeffersonrabb) [inline assets/src/settings/index.js]: Do we actually need a Settings page? And if so, couldn't it just be good old-fashioned PHP-rendered output? Seems like having a React component with a corresponding API endpoint is overkill.

File: dependency-includer-script.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
+<?php
+/**
+ * Dependency includer script.
+ * TODO: use Composer autoloading instead of this.
+ *
+ * @package Newspack
+ */
+
+if ( ! class_exists( '\NewspackContentConverter\Config' ) ) {
```
Comment (jeffersonrabb) [inline dependency-includer-script.php]: Refresh my memory why you took this approach, as opposed  to simply including all of the `include_once`s in the main plugin file?

File: lib/class-converter.php [hunk 3]
```diff
@@ -1,93 +1,361 @@
...
+				'wp-rich-text',
+			],
+			filemtime( plugin_dir_path( __FILE__ ) . '../assets/dist/main.js' ),
+			false
+		);
+
+		wp_enqueue_style(
+			'newspack-content-converter-script',
+			plugins_url( '../assets/dist/main.css', __FILE__ ),
+			[],
```
[the remark this answers is not in the document]
Comment (iuravic) [reply lib/class-converter.php]: Ah, gocha! You were talking about the styles, sorry. Now including style for `wp-components`.

File: assets/src/patchers/index.js [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	};
+
+	/*
+	 * render().
+	 */
+	render() {
+		const { isPatchingOngoing, queuedBatchesPatchingCsv, maxBatchPatching, patchingBatchSize, queuedEntries } = this.state;
+
+		// Non loaded state.
+		if ( null == isPatchingOngoing ) {
```
[the remark this answers is not in the document]
Comment (iuravic) [reply assets/src/patchers/index.js]: You're absolutely right. We don't need the `null` case. I've completely removed it here, as well as for `isConversionOngoing`, and all works well without it.

File: dependency-includer-script.php (hunk 2 above)
Comment (iuravic) [reply dependency-includer-script.php]: Great one. Thanks for the suggestion. Updated, removed all the `if`s.

File: assets/src/index.js [hunk 5]
```diff
@@ -7,18 +7,41 @@ import { render } from '@wordpress/element';
  * Internal dependencies.
  */
 import ContentConverter from './content-converter';
+import ContentRepatcher from './content-repatcher';
+import Settings from './settings';
+import Conversion from './conversion';
+import Patchers from './patchers';
 import * as serviceWorker from './service-worker';
```
[the remark this answers is not in the document]
Comment (iuravic) [reply assets/src/index.js]: Now removed the `service-worker/index.js` and its use here.

**Comment by iuravic** [comment]
@jeffersonrabb updated all the 2nd round comments, ready for re-re-review.

**Pull request merged**